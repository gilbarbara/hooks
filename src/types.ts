import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
