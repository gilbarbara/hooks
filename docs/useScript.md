# useScript

Dynamically loads a script by creating a `<script>` tag and appending it to the `document.body`.  
This hook is useful for integrating third-party scripts like APIs or analytics libraries.

## Usage

```tsx
import { useScript } from '@gilbarbara/hooks';

function Component() {
  const [loaded, error] = useScript(
    'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    'google-maps'
  );

  return (
    <div>
      {loaded && <div id="maps"/>}
      {error && <span>error</span>}
    </div>
  );
}
```

**Custom Attributes**
  
```tsx
import { useScript } from '@gilbarbara/hooks';

function Component() {
  const [loaded] = useScript('https://example.com/script.js', {
    id: 'example-script',
    async: true,
    type: 'module',
  });

  return <div>{loaded ? 'Script is ready!' : 'Loading...'}</div>;
}
```

## Reference

```typescript
interface UseScriptOptions {
  async?: boolean;
  defer?: boolean;
  id?: string;
  type?: string;
}

type UseScriptResult = [loaded: boolean, error: boolean];

useScript(src: string, idOrOptions: string | UseScriptOptions = {}): UseScriptResult;
```
