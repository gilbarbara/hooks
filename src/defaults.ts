import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
