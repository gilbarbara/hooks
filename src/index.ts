export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
