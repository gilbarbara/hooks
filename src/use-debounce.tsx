import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
