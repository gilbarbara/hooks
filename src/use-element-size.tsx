import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
