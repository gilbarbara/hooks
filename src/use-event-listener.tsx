import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
