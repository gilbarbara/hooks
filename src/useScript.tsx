import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { canUseDOM, isString, off, on } from './utils';

interface UseScriptOptions {
  async?: boolean;
  defer?: boolean;
  id?: string;
  type?: string;
}

export type UseScriptResult = [loaded: boolean, error: boolean];

export function useScript(
  src: string,
  idOrOptions: string | UseScriptOptions = {},
): [loaded: boolean, error: boolean] {
  const options = useMemo(
    () => (isString(idOrOptions) ? { id: idOrOptions } : idOrOptions),
    [idOrOptions],
  );
  const script = useRef<HTMLScriptElement>();
  const [state, setState] = useState({
    loaded: false,
    error: false,
  });

  const onLoad = useCallback(() => {
    setState({
      loaded: true,
      error: false,
    });
  }, []);

  const onError = useCallback(() => {
    if (script.current) {
      script.current.remove();
    }

    setState({
      loaded: false,
      error: true,
    });
  }, []);

  useEffect(
    () => {
      if (!canUseDOM() || script.current) {
        return undefined;
      }

      const element = document.createElement('script');

      element.async = options.async ?? true;
      element.defer = options.defer ?? false;
      element.type = options.type || 'text/javascript';
      element.id = options.id || src;
      element.src = src;

      script.current = element;

      const { current } = script;

      on(current, 'load', onLoad);
      on(current, 'error', onError);

      // Add script to document body
      document.body.appendChild(current);

      return () => {
        off(current, 'load', onLoad);
        off(current, 'error', onError);
      };
    },
    [onError, onLoad, options, src], // Only re-run effect if script src changes
  );

  return [state.loaded, state.error];
}
