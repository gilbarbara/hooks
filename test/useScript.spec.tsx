import * as React from 'react';
import { act, render } from '@testing-library/react';

import { useScript } from '../src/useScript';

function Component() {
  const [loaded, error] = useScript(
    'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
  );

  return (
    <div>
      {loaded && <span>loaded</span>}
      {error && <span>error</span>}
    </div>
  );
}

describe('useScript', () => {
  afterEach(() => {
    const script = document.body.querySelector('script');

    if (script) {
      script.remove();
    }
  });

  it('should insert the script and load', async () => {
    const { container } = render(<Component />);
    const script = document.body.querySelector('script');

    expect(script).toHaveAttribute(
      'src',
      'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    );

    act(() => {
      script?.dispatchEvent(new Event('load'));
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should insert the script and fail', async () => {
    const { container } = render(<Component />);
    const script = document.body.querySelector('script');

    expect(script).toHaveAttribute(
      'src',
      'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    );

    act(() => {
      script?.dispatchEvent(new Event('error'));
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
