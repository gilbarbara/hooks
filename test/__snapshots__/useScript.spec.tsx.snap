// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useScript should insert the script and fail 1`] = `
<div>
  <span>
    error
  </span>
</div>
`;

exports[`useScript should insert the script and load 1`] = `
<div>
  <span>
    loaded
  </span>
</div>
`;
